allow to forward line using universal argument

d12frosted opened the pull request:
Hey,

This pretty useful for jumping over several lines or selecting several lines. For example, `C-2 C-a` will move to the beginning of code (line) in the next line. It also allows using `C-- C-2 C-a` to jump to previous line.

alezost commented on mwim.el:
```diff
@@ -124,38 +124,63 @@ to the end of line."
+  (interactive "^p")
+  (setq arg (or arg 1))
+  (when (/= arg 1)
+    (let ((line-move-visual nil))
```
IIUC, setting `line-move-visual` doesn't affect `forward-line`.  It
changes only the behaviour of `line-move`, so there is no need to set it
here.

BTW, this can simply be: `(let (line-move-visual) ...)`

alezost commented on mwim.el:
```diff
@@ -124,38 +124,63 @@ to the end of line."
+(defun mwim-forward-line (arg)
+  "Move ARG - 1 lines forward (backward if ARG is negative).
+If ARG is nil or 1 do nothing."
+  (interactive "^p")
```
Since this function is not for interactive use, it shouldn't be
interactive.

alezost commented on mwim.el:
```diff
@@ -124,38 +124,63 @@ to the end of line."
+If ARG is not nil or 1, move forward ARG - 1 lines first. If
+point reaches the beginning or end of the buffer, stop there."
+  (interactive "^p")
+  (call-interactively #'mwim-forward-line)
```
This is not needed, just: `(mwim-forward-line arg)`.

In general, `call-interactively` is needed very rarely, and (IMHO) most
of the times it is a sign of a bad code :-)

d12frosted replied:
Hm. Didn't know. I thought that it's a good thing to call interactive functions using `call-interactively`.

alezost commented:
Hello, and sorry for the delay.  Using prefix argument is a good idea,
thanks!  However, I don't really like how you implemented it.  I added a
couple of notes to your code.

I find it strange, that you want to use `2` to move to the next line.  I
think it would be better to have a usual meaning for the prefix
argument, i.e.: when you specify `1`, you move one line forward; when it
is `-2`, you move 2 lines backward.

So I came up with another solution:
https://github.com/alezost/mwim.el/commit/c1b413f597ad8c0ad8a820f38d47bf25283d8b6e

WDYT?  If you don't have any comments on it, I'm going to commit it to
master.

BTW, you wrote that specifying `-2` will move to the previous line but
this is not true (did you check it?).  With your code, using `0` prefix
will do it.

Sorry, if this message sounds a bit harsh, I didn't meant it.  I really
like your idea, I just want to write it in different way, thanks!

alezost commented on mwim.el:
```diff
@@ -124,38 +124,63 @@ to the end of line."
+  (when (/= arg 1)
+    (let ((line-move-visual nil))
+      (forward-line (1- arg)))))
+
```
After all, I would write this function like this:

``` elisp
(defun mwim-forward-line (&optional arg)
  "Move ARG - 1 lines forward (backward if ARG is negative).
If ARG is nil or 1 do nothing."
  (when (and arg (/= arg 1))
    (forward-line (1- arg))))
```

But I think this is a strange function: it is used to "forward line",
but not when ARG is 1 (why?), and when you need to go to the next line
you need to give ARG=2. :confused:

d12frosted replied:
This is because when you use `p` option of `interactive`, default value of `arg` is `1`. But I just realised that it's possible to use `P` in combination with `numberp` check. So yeah, since I also don't like using `2` to jump forward, I am getting rid of this function.

d12frosted commented:
Thanks for the review. I've updated my commit. In short:
- I removed redundant `mwim-forward-line` function
- Now line is forwarded by `arg` instead of `arg - 1` and only when `arg` is number and non-zero (otherwise calling `C-0 C-a` twice doesn't switch position).

alezost commented on mwim.el:
```diff
@@ -125,37 +125,61 @@ to the end of line."
+
+If ARG is not nil, move forward ARG lines first. If point reaches
+the beginning or end of the buffer, stop there."
+  (interactive "^P")
```
I think "P" does not suit here: for example, when I use `M--` (or `C--`)
prefix, I want to move one line backward; or `C-u` should move 4 lines
forward – these are the standard meanings for prefixes.  As you noted,
"p" is also not good.  That's why I prefer my solution.  Did you look at
it?

d12frosted replied:
Oh sorry, somehow I missed https://github.com/alezost/mwim.el/pull/2#issuecomment-230110811 😳. Let me take a look.

alezost commented:
I think handling `0` prefix does not worth it, as no one will ever use
it.  I mean `C-0 C-a` is the same as `C-a` so why anyone would use
`C-0`?  But this is a really nice catch!  I didn't think about it.

d12frosted commented:
@alezost yeah, I also think that it's not critical to handle `C-0 C-a`, but just for fun I tried it and… it didn't work :D So for the sake of consistency I decided to handle it. 

> Sorry, if this message sounds a bit harsh, I didn't meant it. I really
> like your idea, I just want to write it in different way, thanks!

No offence is taken 😸 I do understand your intention. And since I am far from being expert in Emacs Lisp I really appreciate your comments. Plus you are OK with having such functionality, so I am double-happy 💃 

And I do like your solution and find it better than mine. The only problem, it seems that it doesn't handle `0` arg. But probably this is not an issue.

Also, very nice trickery about `interactive`. Never seen such usage (when you pass a function). Though just re-read it's description and… I have better understanding of how `interactive` actually works. Thanks for the lesson 😸 

P. S. I am sorry for being spammy today.

alezost commented:
OK, so I committed it.

```
And I do like your solution and find it better than mine. The only
problem, it seems that it doesn't handle 0 arg. But probably this is
not an issue.
```

Eventually I followed your suggestion to check for `0` argument, thanks!

In the latest commit (e93861b06053bbb1872023deb8bb1fb28d91a2bc) I also
added a macro to generate `mwim-{beginning/end}-of-...` commands to
avoid the code duplicating (I should have done this in the very
beginning :blush:).  I hope it didn't break anything.

```
Also, very nice trickery about interactive. Never seen such usage
(when you pass a function).
```

This is actually a very common practice.  It can be met in many
packages.  Using a string in `interactive` form is rather powerful, but
as you can see sometimes we need more.

```
P. S. I am sorry for being spammy today.
```

Heh, no problem at all!  Often I also feel that I'm too verbose :smile:

Thanks again for this pull request!

d12frosted commented:
Hooray! This is great!

> Eventually I followed your suggestion to check for 0 argument, thanks!

💃 

> In the latest commit (e93861b) I also
> added a macro to generate mwim-{beginning/end}-of-... commands to
> avoid the code duplicating (I should have done this in the very
> beginning 😊). I hope it didn't break anything.

Wow, this looks nice. 💯 

> This is actually a very common practice. It can be met in many
> packages. Using a string in interactive form is rather powerful, but
> as you can see sometimes we need more.

Hm. Probably I never seen this before because I am pretty new to writing Emacs Lisp. But indeed, it's great that `interactive` allows passing a function to be better at handling arguments. Emacs as ecosystem (with it's ultimate approach to extensibility) amazes me every day 😸 

> Thanks again for this pull request!

You're welcome. And thanks for resolving my feature request 😸
